File: euler/03/problem3.cpp
#include "problem3.h"
#include <math.h>
#include <iostream>
#include <vector>
#include <chrono>

using namespace std::chrono;
// g++ problem3.cpp problem3.h --std=c++11 -o problem3
// ./problem3

bool is_prime(int n) {
  if (n == 2 || n == 3) {
    return true;
  }
  if (n % 2 == 0 || n < 2) {
    return false;
  }
  for (int i = 3; i <= (floor(pow(n, 0.5)) + 1); i += 2) {
    if (n % i == 0) {
      return false;
    }
  }
  return true;
}
int main() {
  high_resolution_clock::time_point start =
      high_resolution_clock::now();

  long top = 600851475143;
  std::vector<int> factors;
  for (int i = 1; i < top; i++) {
    if (top % i == 0 && is_prime(i)) {
      factors.push_back(i);

      long topcopy = top;
      for (int j : factors) {
        topcopy /= j;
      }
      if (topcopy == 1) {
        break;
      }
    }
  }
  high_resolution_clock::time_point end = high_resolution_clock::now();
  duration<double> span = duration_cast<duration<double>>(end - start);

  std::cout << factors[factors.size() - 1] << std::endl;
  std::cout << span.count() * 1000 << " ms" << std::endl;
}

File: euler/03/problem3.py
# python3.5 problem3.py
import time

def is_prime(n): # Definitely not copied from the internet
    # I've been exposed cause I don't write code like that lol
    if n==2 or n==3: return True
    if n%2==0 or n<2: return False
    for i in range(3,int(n**0.5)+1,2):   # only odd numbers
        if n%i==0:
            return False    
    return True

start = time.time()

top = 600851475143
factors = []
for i in range(1, top, 1):
  if top % i == 0 and is_prime(i):
    factors.append(i)

    topcopy = top # why does this work wtf
    for j in factors:
      topcopy /= j
    if topcopy == 1:
      break  

print(factors[-1]) # Most recent (and largest) prime factor
print(str((time.time() - start) * 1000) + " ms") # hate you python
